File: 3233-find-the-count-of-numbers-which-are-not-special/3233-find-the-count-of-numbers-which-are-not-special.py
class Solution(object):
    def nonSpecialCount(self, l, r):
        """
        :type l: int
        :type r: int
        :rtype: int
        """
        prime = [True] * (int(math.sqrt(r)) + 5)
        v = []
        cnt = 0
        self.SieveOfEratosthenes(len(prime) - 1, prime, v)
        for x in v:
            val = x * x
            if l <= val <= r:
                cnt += 1
        return r - l + 1 - cnt
    def SieveOfEratosthenes(self, n, prime, v):
        prime[0] = prime[1] = False
        for p in range(2, int(math.sqrt(n)) + 1):
            if prime[p]:
                for i in range(p * p, n + 1, p):
                    prime[i] = False
        for p in range(2, n + 1):
            if prime[p]:
                v.append(p)

File: 3233-find-the-count-of-numbers-which-are-not-special/3233-find-the-count-of-numbers-which-are-not-special.java
class Solution {
    private static HashSet<Integer> set = new HashSet<>() {
        {
            boolean[] flag = new boolean[31608];
            for (int i = 2; i < 31608; i++) {
                if (!flag[i]) {
                    add(i);
                    for (int j = 2 * i; j < 31608; j += i) {
                        flag[j] = true;
                    }
                }
            }
        }
    };

    public int nonSpecialCount(int l, int r) {
        int count = r - l + 1;
        for (int i : set) {
            count -= i * i < l || i * i > r ? 0 : 1;
        }
        return count;
    }
}